### ISSUE TITLE
Loadbalancer NLB Target Group health checks failing since upgrade to v1.16.0

### ISSUE THREAD

**Issue by okamosy** [cilium/cilium#34093]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

higher than v1.16.0 and lower than v1.17.0

### What happened?

I am running a cluster in AWS EKS with three nodes using the Ingress Controller to manage a network load balancer within AWS. After upgrading to 1.16.0, all health checks on the target groups began to fail. Strangely, replacing a node would allow the checks on port 80 for the new instance to return healthy, at least until the Cilium DS is restarted.

While troubleshooting the issue, I noticed that `cilium status` would periodically return the following error:
```
controller node-neighbor-link is failing since 4s (2x): unable to determine next hop IPv4 address for eth1 (<node_ip>): remote node IP is non-routable
unable to determine next hop IPv4 address for eth2 (<node_ip>): remote node IP is non-routable
unable to determine next hop IPv4 address for eth3 (<node_ip>): remote node IP is non-routable
```

I exec'd into the pod reporting this and ran `cilium-dbg status`  and only thing of note was this line:
```
Modules Health: Stopped(0) Degraded(3) OK(148)
```

Running `cilium-dbg status --verbose` showed that node-manager was degraded with the message `Failed node neighbor link update`. I have not been able to determine if the two issues are connected, but I have stripped down our values.yaml to the bare minimum of settings in order to see if our configuration is at fault, but have not been able to restore the health checks.

### How can we reproduce the issue?

1. Install Cilium with the following values.yaml file:

```yaml
egressMasqueradeInterfaces: eth0
eni:
  awsReleaseExcessIPs: true
  enabled: true
envoy:
  enabled: true
ingressController:
  enableProxyProtocol: false
  enabled: true
  loadbalancerMode: shared
  service:
    annotations:
      service.beta.kubernetes.io/aws-load-balancer-alpn-policy: HTTP2Preferred
      service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp
      service.beta.kubernetes.io/aws-load-balancer-internal: "true"
      service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: instance
      service.beta.kubernetes.io/aws-load-balancer-ssl-cert: <ssl-cert-arn>
      service.beta.kubernetes.io/aws-load-balancer-ssl-negotiation-policy: ELBSecurityPolicy-TLS13-1-2-2001-06
      service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
      service.beta.kubernetes.io/aws-load-balancer-target-group-attributes: preserve_client_ip.enabled=false
      service.beta.kubernetes.io/aws-load-balancer-type: nlb
ipam:
  mode: eni
ipv4NativeRoutingCIDR: <CIDR>
k8sServiceHost: <eks-endpoint>
k8sServicePort: 443
kubeProxyReplacement: true
routingMode: native
```

### Cilium Version

v1.16.0

### Kernel Version

5.10.219-208.866.amzn2.x86_64

### Kubernetes Version

v1.29.4-eks-036c24b

### Regression

v1.15.6

### Cilium Users Document

- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ridwansswnto** [comment]
Have you tried changing the policy enforcement mode? currently I'm using the policy enforcement `default` the instance is not accessible for its NodePort, I tried changing it to `never` and the instance is accessible for its NodePort again and the NLB is healthy.
```
enable-policy                                     never
```

**Comment by dvb-simp** [comment]
hi @ridwansswnto, 
I tested your solution, but it did not change the behavior, the targets are still `Unhealthy`. Can you elaborate a bit more on your current setup? Thanks in advance

**Comment by dvb-simp** [comment]
I think I might have found the solution to this problem @okamosy ; 
`bpf.enableTCX=false` should be set to the values.yaml file when the kernel version is lower than 6.6.
This was stated in the [https://docs.cilium.io/en/stable/operations/upgrade/](upgrade notes). 

In our case, the kernel version was 6.1.. I needed to restart the nodes to have this setting propagated. 
Will do some more investigation on this.

**Comment by okamosy** [comment]
> I think I might have found the solution to this problem @okamosy ; `bpf.enableTCX=false` should be set to the values.yaml file when the kernel version is lower than 6.6. This was stated in the upgrade-notes [https://docs.cilium.io/en/stable/operations/upgrade/](url).
> 
> In our case, the kernel version was 6.1.. I needed to restart the nodes to have this setting propagated. Will do some more investigation on this.

Thanks for this. I've tried it, and I still get the error about `Failed node neighbor link update` between nodes. I will run some more tests in the morning to see if I can just ignore those messages, but I think I've still running into some issue.

**Comment by dvb-simp** [comment]
> > I think I might have found the solution to this problem @okamosy ; `bpf.enableTCX=false` should be set to the values.yaml file when the kernel version is lower than 6.6. This was stated in the upgrade-notes [https://docs.cilium.io/en/stable/operations/upgrade/](url).
> > In our case, the kernel version was 6.1.. I needed to restart the nodes to have this setting propagated. Will do some more investigation on this.
> 
> Thanks for this. I've tried it, and I still get the error about `Failed node neighbor link update` between nodes. I will run some more tests in the morning to see if I can just ignore those messages, but I think I've still running into some issue.

You're right! I see the neighbor events from the `cilium status --verbose` as well, eventhough the targets from the NLB are healthy. After Restarting the Cilium agent Daemonset, all become targets Unhealthy again. This needs more investigation as this is not a suitable way of having the upgraded cilium instances.

**Comment by ridwansswnto** [comment]
> 

> hi @ridwansswnto, I tested your solution, but it did not change the behavior, the targets are still `Unhealthy`. Can you elaborate a bit more on your current setup?
> 
> Our setup is: with the spec mentioned by @rmdvb Upgrade from cilium v1.15.7 to v1.16.0, and the preflights are healthy. Cilium 1.16.0, Bottlerocket v1.20.4 and kubernetes v1.28
> 
> ```
> # sanitized values.yaml file
> cilium:
>   bpf:
>     masquerade: true
>   kubeProxyReplacement: "true"
>   ipam:
>     operator:
>       clusterPoolIPv4PodCIDRList: [<cidrlist>]
>   cni:
>     install: true
>     chainingMode: none
>   envoyConfig:
>     enabled: false
>   ingressController:
>     enabled: true
>     loadbalancerMode: "shared"
>     service:
>       name: cilium-ingress
>       annotations:
>         service.beta.kubernetes.io/aws-load-balancer-type: external
>         service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: instance
>         service.beta.kubernetes.io/aws-load-balancer-scheme: internal
>         service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-protocol: tcp
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-port: traffic-port
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-healthy-threshold: "2"
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-unhealthy-threshold: "2"
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-timeout: "5"
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-interval: "5"
>         service.beta.kubernetes.io/aws-load-balancer-attributes: "load_balancing.cross_zone.enabled=false"
>         service.beta.kubernetes.io/load-balancer-source-ranges: "<source-ranges>"
>         service.beta.kubernetes.io/aws-load-balancer-target-group-attributes: "deregistration_delay.timeout_seconds=60,deregistration_delay.connection_termination.enabled=true,preserve_client_ip.enabled=true"
>         service.beta.kubernetes.io/aws-load-balancer-subnets: "<lb-subnets>"
>       type: LoadBalancer
>       insecureNodePort: 30080
>       secureNodePort : 30443
>   eni:
>     enabled: false
>     updateEC2AdapterLimitViaAPI: false
>   policyEnforcementMode: "always"
>   policyAuditMode: true
>   envoy:
>     enabled: false
>   tunnelProtocol: "vxlan"
> ```

> hi @ridwansswnto, I tested your solution, but it did not change the behavior, the targets are still `Unhealthy`. Can you elaborate a bit more on your current setup?
> 
> Our setup is: with the spec mentioned by @rmdvb Upgrade from cilium v1.15.7 to v1.16.0, and the preflights are healthy. Cilium 1.16.0, Bottlerocket v1.20.4 and kubernetes v1.28
> 
> ```
> # sanitized values.yaml file
> cilium:
>   bpf:
>     masquerade: true
>   kubeProxyReplacement: "true"
>   ipam:
>     operator:
>       clusterPoolIPv4PodCIDRList: [<cidrlist>]
>   cni:
>     install: true
>     chainingMode: none
>   envoyConfig:
>     enabled: false
>   ingressController:
>     enabled: true
>     loadbalancerMode: "shared"
>     service:
>       name: cilium-ingress
>       annotations:
>         service.beta.kubernetes.io/aws-load-balancer-type: external
>         service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: instance
>         service.beta.kubernetes.io/aws-load-balancer-scheme: internal
>         service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-protocol: tcp
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-port: traffic-port
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-healthy-threshold: "2"
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-unhealthy-threshold: "2"
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-timeout: "5"
>         service.beta.kubernetes.io/aws-load-balancer-healthcheck-interval: "5"
>         service.beta.kubernetes.io/aws-load-balancer-attributes: "load_balancing.cross_zone.enabled=false"
>         service.beta.kubernetes.io/load-balancer-source-ranges: "<source-ranges>"
>         service.beta.kubernetes.io/aws-load-balancer-target-group-attributes: "deregistration_delay.timeout_seconds=60,deregistration_delay.connection_termination.enabled=true,preserve_client_ip.enabled=true"
>         service.beta.kubernetes.io/aws-load-balancer-subnets: "<lb-subnets>"
>       type: LoadBalancer
>       insecureNodePort: 30080
>       secureNodePort : 30443
>   eni:
>     enabled: false
>     updateEC2AdapterLimitViaAPI: false
>   policyEnforcementMode: "always"
>   policyAuditMode: true
>   envoy:
>     enabled: false
>   tunnelProtocol: "vxlan"
> ```

My bad, I keep experiencing the same issues.

This is my setup:
eks 1.29, bottlerocket 1.20.5-a3e8bda1, cilium 1.16.0.

values.yaml
```
agent: true
k8sServiceHost: ""
k8sServicePort: ""
debug:
  enabled: true
  verbose: "flow policy envoy datapath"
cluster:
  name: ""
upgradeCompatibility: "1.8"
ipam:
  mode: "eni"
kubeProxyReplacement: "true"
hubble:
  relay:
    enabled: true
  ui:
    enabled: true
eni:
  awsEnablePrefixDelegation: true
  awsReleaseExcessIPs: true
  ec2APIEndpoint: ""
  enabled: true
  eniTags: {}
  gcInterval: ""
  gcTags: {}
  iamRole: ""
  instanceTagsFilter: []
  subnetIDsFilter: [""]
  subnetTagsFilter: []
  updateEC2AdapterLimitViaAPI: true
ingressController:
  enabled: true
  loadbalancerMode: shared
  enforceHttps: true
  enableProxyProtocol: false
  ingressLBAnnotationPrefixes: ['lbipam.cilium.io', 'nodeipam.cilium.io', service.beta.kubernetes.io, 'service.kubernetes.io', 'cloud.google.com']
  secretsNamespace:
    create: true
    name: cilium-secrets
    sync: true
  hostNetwork:
    enabled: false
  service:
    type: LoadBalancer
    insecureNodePort: 32480
    secureNodePort: 32443
    name: cilium-ingress
    annotations:
      service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
      service.beta.kubernetes.io/aws-load-balancer-internal: "true"
      service.beta.kubernetes.io/aws-load-balancer-subnets: ""
```

After several attempts like rolling update instances etc, I found something:

This is the bpf lb status when NLB Healthcheck fails:
```
$ kubectl -n kube-system exec -ti ds/cilium -- cilium bpf lb list
10.12.12.88:32182 (0)     0.0.0.0:0 (102) (0) [NodePort]
10.12.13.70:32182 (0)     0.0.0.0:0 (103) (0) [NodePort]
```

This is the bpf lb status when NLB Heatcheck is successful:
```
$ kubectl -n kube-system exec -ti ds/cilium -- cilium bpf lb list
10.12.13.95:32480 (0)     0.0.0.0:0 (21) (0) [NodePort, l7-load-balancer] (L7LB Proxy Port: 10969)
10.12.13.95:32443 (0)     0.0.0.0:0 (23) (0) [NodePort, l7-load-balancer] (L7LB Proxy Port: 10969)
```

The interesting thing is that if I update something part of `ingressController` WITHOUT restarting `cilium-operator` & `cilium-agent` the bpf lb will show descriptions like `l7-load-balancer` and HC in NLB success, once I restart they are gone and the NLB is unhealthy.

This is the log for healthy NLB in cilium-agent:
```
time="2024-08-01T14:45:41Z" level=debug msg="Upserting service" backends="[]" l7LBProxyPort=10969 loadBalancerSourceRanges="[]" serviceIP="{10.12.13.95 {TCP 32480} 0}" serviceName=cilium-ingress serviceNamespace=kube-system sessionAffinity=false sessionAffinityTimeout=0 subsys=service svcExtTrafficPolicy=Cluster svcHealthCheckNodePort=0 svcIntTrafficPolicy=Cluster svcType=NodePort
time="2024-08-01T14:45:41Z" level=debug msg="Acquired service ID" backends="[]" l7LBProxyPort=10969 loadBalancerSourceRanges="[]" serviceID=21 serviceIP="{10.12.13.95 {TCP 32480} 0}" serviceName=cilium-ingress serviceNamespace=kube-system sessionAffinity=false sessionAffinityTimeout=0 subsys=service svcExtTrafficPolicy=Cluster svcHealthCheckNodePort=0 svcIntTrafficPolicy=Cluster svcType=NodePort

```

This is the log after `cilium-operator` & `cilium-agent` restart:
```
time="2024-08-01T15:04:42Z" level=debug msg="Upserting service" backends="[]" l7LBProxyPort=0 loadBalancerSourceRanges="[]" serviceIP="{10.12.13.95 {NONE 32480} 0}" serviceName=cilium-ingress serviceNamespace=kube-system sessionAffinity=false sessionAffinityTimeout=0 subsys=service svcExtTrafficPolicy=Cluster svcHealthCheckNodePort=0 svcIntTrafficPolicy=Cluster svcType=NodePort
time="2024-08-01T15:04:42Z" level=debug msg="Acquired service ID" backends="[]" l7LBProxyPort=0 loadBalancerSourceRanges="[]" serviceID=21 serviceIP="{10.12.13.95 {NONE 32480} 0}" serviceName=cilium-ingress serviceNamespace=kube-system sessionAffinity=false sessionAffinityTimeout=0 subsys=service svcExtTrafficPolicy=Cluster svcHealthCheckNodePort=0 svcIntTrafficPolicy=Cluster svcType=NodePort
```

**Comment by okamosy** [comment]
> The `Failed node neighbor link update` and
> 
> ```
> controller node-neighbor-link is failing since 4s (2x): unable to determine next hop IPv4 address for eth1 (<node_ip>): remote node IP is non-routable
> unable to determine next hop IPv4 address for eth2 (<node_ip>): remote node IP is non-routable
> unable to determine next hop IPv4 address for eth3 (<node_ip>): remote node IP is non-routable
> ```
> 
> Messages would indicate that there are no route which can be used to reach the remote node. Could you inspect the routing table of the nodes where this error occurs or provide a sysdump which should contain this info as well.

I've attached the sys dump from a clean build of an EKS cluster and Cilium 1.16.0
[cilum-sysdump-clean-1.16-install.zip](https://github.com/user-attachments/files/16472405/cilum-sysdump-clean-1.16-install.zip)

Additionally, I've disabled the Ingress Controller to try and make sure it wasn't getting in the way of something else.

**Comment by mattcharlton** [comment]
@ridwansswnto @dvb-simp @rmdvb are you folks still experiencing the same issues with this upgrade? We're still struggling.

**Comment by dvb-simp** [comment]
Hi @mattcharlton, we're still seeing this issue and it's currently blocking us to upgrade to cilium v1.16.0... 
It seems that the cilium ingress LB NodePorts aren't accepting any traffic after a restart of the cilium agent, but we can't find any logs or events that triggers this behaviour.

**Comment by devang704** [comment]
We also run into a similar issue in cilium v1.16.0 upgrade. After update some parameter in cilium, the cilium pods are restarted and after that Health check failing for all target groups due to the node Port stop responding.

**Comment by tommasopozzetti** [comment]
We are running into this exact same issue as well with 1.16.0. We do not see the neighbor link issues but we have the exact behavior described by @ridwansswnto above.
Specifically we can see this very well with the ingress controller shared LB service. If we delete it and reapply it, running `kubectl -n kube-system exec -ti ds/cilium -- cilium bpf lb list` shows the entries with also option `l7-load-balancer` and everything works correctly. If we rollout restart cilium daemonset (we are using embedded envoy) without making any change, the endpoint start giving `connection refused` and running `cilium bpf lb list` shows the endpoint are no longer listed with the `l7-load-balancer` option.

**Comment by tommasopozzetti** [comment]
I just realized that 1.16.1 had come out earlier today and tested it and I cannot reproduce the issue with 1.16.1 🎉 
Combed through the release notes and it seems like https://github.com/cilium/cilium/pull/34131 might be what fixed the issue for me

**Comment by DevlinMS** [comment]
I was having the same issue on 1.16.0 and 1.16.1. I was able to resolve the issue by defining a `CiliumNodeConfig` resource for each node that had different network devices available. 

More specifically, my helm install included `devices: "eth+ br+ eno+"`, which I switched out for 
```
apiVersion: cilium.io/v2
kind: CiliumNodeConfig
metadata:
  namespace: kube-system
  name: bridged-device
spec:
  nodeSelector:
    matchLabels:
      device: "bridge"
  defaults:
    devices: "br+"
```

**Comment by mattcharlton** [comment]
We've tried 1.16.1 - same problem with no change.  Tried an in-place cluster upgrade and a new cluster build.  

Load balancer shows the 443 listeners unhealthy, the 80s are fine - looking at the bpf lb list in cilum the port 80 has that l7-load-balancer label/option, but its missing from the unhealthy 443 one.  Tried restarted the DSs to no avail.  

Something still doesn't seem right within cilium itself here.

**Comment by BenoitKnecht** [comment]
I'm also affected by this issue, which forced me to rollback from 1.16.1 to 1.15.8.

On a cluster without BGP control plane setup, I didn't see any of the `unable to determine next hop IPv4 address for eth2 (<node_ip>): remote node IP is non-routable` errors, but with a `CiliumBGPPeeringPolicy` in place, Cilium 1.16.1 keeps logging those errors.

This issue may not be very widespread because it also requires the node to have more than one network interface. For instance on a node with

```console
$ ip route
default via 10.0.0.1 dev eth0
172.16.0.0/12 via 172.16.0.1 dev eth1
```

Cilium seems to try and find a route to a remote node with IP address `10.0.0.7` _through each interface explicitly_; it finds a path though `eth0` of course, as it has the default route, but it doesn't find it through `eth1` and therefore logs

```
node-neighbor-link-updater is failing since 4s (1x): unable to determine next hop IPv4 address for eth1 (10.0.0.7): remote node IP is non-routable
```

Looking at the code, it looks like Cilium is going through each interface it knows about and doing the equivalent of

```console
$ ip route get 10.0.0.7 dev eth1 fib
```

which fails if there's no path through that specific interface.

So it would seem like without the BGP configuration, Cilium isn't trying to find out about those routes for some reason, or at least not on the same set of interfaces. This is also why @DevlinMS's workaround gets rid of these errors, because now Cilium is only aware of the interface with the default route. Unfortunately in my case, I also need some traffic to go through the _other_ interface, so if I apply the workaround I don't get the error anymore, but some traffic ends up being blackholed.

Has anyone being digging deeper into this issue? It seems like a pretty serious regression to me, and it's definitely preventing me from upgrading to 1.16+ for the time being.

**Comment by jmicceri** [comment]
We are also running into this issue, we started out on `1.16.1` and we don't really have the option of downgrading to `1.15.x`. Initially I couldn't get any of the targets to be healthy but once I set `externalTrafficPolicy = "Local"` they became healthy right away. However as others have mentioned, as soon as the agent restarts or a fresh node comes up, the targets start failing health checks and never recover.

Are there any other work arounds or fixes in progress for this one?

**Comment by andjhop** [comment]
It's difficult to know if all the reported observations are related. I've been able to reproduce the issue with unhealthy nodeports after restarting the agent daemonset on versions of Cilium as far back as v1.14.0. Some of the issues being experienced with failing healthchecks on port 443 but not port 80, I believe, may be the the result of #32744 released with v1.16.0. Specifically, a change to the behaviour of the the operator in how it translates ingress resources into CEC resources. This is a likely candidate if your your `CiliumEnvoyConfig` lists only only port 80, for example:
```yaml
apiVersion: cilium.io/v2
kind: CiliumEnvoyConfig
spec:
  services:
  - listener: ""
    name: cilium-ingress
    namespace: kube-system
    ports:
    - 80
```

Previously, port 443 would redirect to the local L7 proxy (Envoy) by default. Now the ingress resource needs to define tls explicitly as in the example here https://kubernetes.io/docs/concepts/services-networking/ingress/#tls:
```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: tls-example-ingress
spec:
  tls:
  - hosts:
      - https-example.foo.com
    secretName: testsecret-tls
  rules:
  - host: https-example.foo.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: service1
            port:
              number: 80
```
